fix(runtime): dynamic slot name change

tanner-reits opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help: https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md -->


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Stencil's runtime will not change the name of a `slot` element when the `name` attribute is bound to a `@Prop()`. This is an issue for both shadow and non-shadow components.

Fixes: #2982 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

The runtime will now correctly update the `name` attribute of slots and pseudo-slots to render the correct content on re-renders. 

## Documentation

https://github.com/ionic-team/stencil-site/pull/1337

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Testing

<!-- Please describe the steps you took to test the changes in this PR. These steps can be programmatic (e.g. unit tests) and/or manual. -->

**Automated Testing**
New e2e tests were added for this use case

**Manual Testing**
1. Spin-up a new Stencil component starter
2. Enable `extras.experimentalSlotFixes` in the Stencil config
3. Update the contents of `my-component.tsx` to the following:
```tsx
import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'my-component',
  styleUrl: 'my-component.css',
  scoped: true,
})
export class MyComponent {
  /**
   * The first name
   */
  @Prop({ reflect: true }) slotName = 'main';

  render() {
    return (
      <div>
        <slot name={this.slotName}></slot>
      </div>
    );
  }
}
```
4. Update the content of `body` in `index.html` to the following:
```html
    <my-component>
      <p slot="main">Hello</p>
      <p slot="secondary">Goodbye</p>
    </my-component>

    <button>Toggle</button>

    <script>
      document.querySelector('button').addEventListener('click', () => {
        document.querySelector('my-component').setAttribute('slot-name', 'secondary');
      });
    </script>
```
5. Run `npm start`
6. After clicking the bottom the contents should not change
7. Install a build of this branch. Run `npm start` again and observe the contents change when clicking the button
8. Repeat while setting `shadow: true` on the component

## Other information

<!-- Any other information that is important to this PR such as screenshots of how a component looks before and after the change. -->

The fix for the `scoped` variation of this issue is placed behind the `extras.experimentalSlotFixes` flag.

rwaskiewicz commented:
One question regarding behavior. Below I have 2 videos - both use the example code that you have in the PR summary. 

The first video uses `shadow: true` in the component. When I hit the button, the "Goodbye" text gets slotted into the shadow dom, and the "Hello" text disappears.
https://github.com/ionic-team/stencil/assets/1930213/f6fc975e-968d-44e7-8f85-06fdfe58e5e1

The first video uses `scoped: true` in the component. When I hit the button, the 'Hello' text is added to the light-ish dom (since we're using scoped, it's all the light dom 😛), and the 'Goodbye' text is slotted into the slot in `my-component`:
https://github.com/ionic-team/stencil/assets/1930213/a10f4397-b171-4f85-adf8-72c8f5d695f9

Is this difference in behavior expected here?

tanner-reits commented:
@rwaskiewicz So turns out the `scoped` variation for this fix relies on another slot fix we put behind `experimentalSlotFixes` (I just didn't realize I had that flag on locally and our e2e tests enable it). So, I updated the code to put this fix behind the flag as well. Updated the PR summary and testing steps to include enabling that flag to see the desired effects. I'll get a docs PR open to list this fix behind that flag as well. Sorry bout that!

christian-bromann reviewed (approved):
👍
